### PR TITLE
Gate initial and last turn autosaves with OptionsDB options.

### PR DESCRIPTION
**Pull request by LGM-Doyle**
This addresses the concern that the autoquit autosave was not gated by an option, raised here https://github.com/freeorion/freeorion/pull/1653#issuecomment-320354748.

It also adds an option to gate the initial turn autosave.

The options are named:
- autosave.initial-turn
- autosave.last-turn

This PR depends on #1665, because they both make changes to the shutdown sequence.

### CODE REVIEW

File: client/human/HumanClientApp.cpp [hunk 1]
```diff
@@ -625,16 +627,29 @@ void HumanClientApp::CancelMultiplayerGameFromLobby()
...
+    DebugLogger() << "Sent save initiate message to server.";
 }
 
 void HumanClientApp::SaveGameCompleted() {
-    DebugLogger() << "HumanClientApp::SaveGameCompleted by server.";
-    m_save_game_in_progress = false;
-    SaveGameCompletedSignal();
+    m_game_saves_in_progress.pop();
+
+    // Either indicate that all save are completed or start the next save.
```
Comment (geoffthemedio) [inline client/human/HumanClientApp.cpp]: `save` -> `saves`

When would there be multiple saves in progress simultaneously?

Comment (LGM-Doyle) [reply client/human/HumanClientApp.cpp]: In the current game if the user saves while the autosave is in progress.

It only matters if the user saves and then immediately exits.  In the current game that means only the first save that completes is kept, the autosave.  The user save is always aborted.  

With PR #1671, the scenario would happen whenever the user saves and then immediately quits.  Although in that case the last turn autosave would be lost.

**Comment by geoffthemedio** [comment]
autosave_options

Seems to work as intended.

The first turn autosave is labelled as being from turn 1, so the in-game description should match that.

There is a missing string for the description in the tooltip.

I'd rename the options "Autosave on first turn" and "Autosave when resigning or closing".

The descriptions should also not just reiterate the name label... so "On first turn of a new game, generate an autosave of the initial game state, before any player actions are taken" and "When resigning a game or closing the application window, generate an autosave" or similar.

File: UI/OptionsWnd.cpp [hunk 2]
```diff
@@ -692,6 +692,8 @@ void OptionsWnd::CompleteConstruction() {
     BoolOption(current_page, 0, "autosave.multiplayer",   UserString("OPTIONS_MULTIPLAYER"));
     IntOption(current_page,  0, "autosave.turns",         UserString("OPTIONS_AUTOSAVE_TURNS_BETWEEN"));
     IntOption(current_page,  0, "autosave.limit",         UserString("OPTIONS_AUTOSAVE_LIMIT"));
+    BoolOption(current_page, 0, "autosave.initial-turn",  UserString("OPTIONS_AUTOSAVE_INITIAL_TURN"));
+    BoolOption(current_page, 0, "autosave.last-turn",     UserString("OPTIONS_AUTOSAVE_LAST_TURN"));
```
Comment (geoffthemedio) [inline UI/OptionsWnd.cpp]: maybe rename these? it's not just an initial turn autosave, it's a before anything happens initial gamestate autosave. It's also not a last turn autosave, it's an autosave when closing or resigning.

File: default/stringtables/en.txt [hunk 3]
```diff
@@ -1638,6 +1638,12 @@ Sets the number of turns between autosaves.
 OPTIONS_DB_AUTOSAVE_LIMIT
 Sets the maximum number of autosave files to keep.
 
+OPTIONS_DB_AUTOSAVE_INTIAL_TURN
+If true, autosave the initial turn before any game play.
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: Prefer "Toggles whether to" rather than "If true, "

File: default/stringtables/en.txt [hunk 4]
```diff
@@ -2267,6 +2273,12 @@ Autosaves limit
 OPTIONS_AUTOSAVE_TURNS_BETWEEN
 Turns between autosaves
 
+OPTIONS_AUTOSAVE_INITIAL_TURN
+Autosave on turn 0 before any game play
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: turn 1 I think?

File: UI/OptionsWnd.cpp (hunk 2 above)
Comment (geoffthemedio) [reply UI/OptionsWnd.cpp]: should there be a distinction between galaxy creation and everything that happens before the start of the first turn?

Comment (LGM-Doyle) [reply UI/OptionsWnd.cpp]: There is a distinction in the server side code, but the programmer can see that the save is generated in the human client FSM `WaitingForGameStart` when the first turn game data is extracted.

I don't think there is a distinction to a player.

I forgot to change the variable names in the non-UI code in `HumanClientApp`, so I have re-written the last commit.

Comment (geoffthemedio) [reply UI/OptionsWnd.cpp]: But the player might see the distinction if/when we make it possible to save and edit galaxy maps (at various stages of being populated with stuff) independently from saved games...

**Comment by geoffthemedio** [comment]
I consistently get crashes from a null pointer exception when starting a multiplayer game with two players. The stack:
````
>	FreeOrion.exe!HumanClientApp::SaveGameCompleted() Line 645	C++
 	FreeOrion.exe!PlayingTurn::react(const SaveGameComplete & msg) Line 898	C++
 	FreeOrion.exe!boost::statechart::simple_state<PlayingTurn,PlayingGame,boost::mpl::list<boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,0>::local_react_impl_non_empty::local_react_impl<boost::mpl::list<boost::statechart::custom_reaction<SaveGameDataRequest>,boost::statechart::custom_reaction<SaveGameComplete>,boost::statechart::custom_reaction<AdvanceTurn>,boost::statechart::custom_reaction<TurnUpdate>,boost::statechart::custom_reaction<TurnEnded>,boost::statechart::custom_reaction<PlayerStatus>,boost::statechart::custom_reaction<DispatchCombatLogs>,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,boost::statechart::simple_state<PlayingTurn,PlayingGame,boost::mpl::list<boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,0> >(boost::statechart::simple_state<PlayingTurn,PlayingGame,boost::mpl::list<boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,0> & stt, const boost::statechart::event_base & evt, const void * eventType) Line 820	C++
 	FreeOrion.exe!boost::statechart::simple_state<PlayingTurn,PlayingGame,boost::mpl::list<boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,0>::react_impl(const boost::statechart::event_base & evt, const void * eventType) Line 489	C++
 	FreeOrion.exe!boost::statechart::state_machine<HumanClientFSM,IntroMenu,std::allocator<void>,boost::statechart::null_exception_translator>::send_event(const boost::statechart::event_base & evt) Line 881	C++
 	FreeOrion.exe!boost::statechart::state_machine<HumanClientFSM,IntroMenu,std::allocator<void>,boost::statechart::null_exception_translator>::process_event(const boost::statechart::event_base & evt) Line 275	C++
 	FreeOrion.exe!HumanClientApp::HandleMessage(Message & msg) Line 886	C++
 	FreeOrion.exe!HumanClientApp::HandleSystemEvents() Line 865	C++
 	GiGi.dll!GG::ModalEventPump::operator()() Line 114	C++
 	GiGiSDL.dll!GG::SDLGUI::Run() Line 883	C++
 	FreeOrion.exe!mainSetupAndRun() Line 258	C++
 	FreeOrion.exe!wmain(int argc, wchar_t * * argv, wchar_t * * envp) Line 104	C++
````
Seems like the non-host player is getting the `SaveGameComplete` message after not having initiated a save and trying to pop an empty stack?

Just putting `if (!m_game_saves_in_progress.empty())` before the stack pop in `HumanClientApp::SaveGameCompleted()` avoids the crash.

**Comment by geoffthemedio** [comment]
I also get an error message in game (no crash) when resigning from a game with no hosts using the hostless branch. It tries to save on exit, but can't because it's not the host. I assume this would also happen if the previous crash didn't in master.

**Pull request merged**